x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/action/EqlSearchResponse.java

```diff
@@ -414,53 +411,30 @@ public void writeTo(StreamOutput out) throws IOException {
             }
         }
 
+        private static final ConstructingObjectParser<EqlSearchResponse.Hits, Void> PARSER =
+            new ConstructingObjectParser<>("eql/search_response_count", true,
+                args -> {
+                    int i = 0;
+                    @SuppressWarnings("unchecked") List<SearchHit> searchHits = (List<SearchHit>) args[i++];
+                    @SuppressWarnings("unchecked") List<Sequence> sequences = (List<Sequence>) args[i++];
+                    @SuppressWarnings("unchecked") List<Count> counts = (List<Count>) args[i++];
+                    TotalHits totalHits = (TotalHits) args[i];
+                    return new EqlSearchResponse.Hits(searchHits, sequences, counts, totalHits);
+                });
+
+        static {
+            PARSER.declareObjectArray(ConstructingObjectParser.optionalConstructorArg(), (p, c) -> SearchHit.fromXContent(p),
+                new ParseField(Fields.EVENTS));
+            PARSER.declareObjectArray(ConstructingObjectParser.optionalConstructorArg(), Sequence.PARSER,
+                new ParseField(Fields.SEQUENCES));
+            PARSER.declareObjectArray(ConstructingObjectParser.optionalConstructorArg(), Count.PARSER,
+                new ParseField(Fields.COUNTS));
+            PARSER.declareObject(ConstructingObjectParser.optionalConstructorArg(), (p, c) -> SearchHits.parseTotalHitsFragment(p),
+                new ParseField(Fields.TOTAL));
+        }
+
         public static Hits fromXContent(XContentParser parser) throws IOException {
-            if (parser.currentToken() != XContentParser.Token.START_OBJECT) {
-                parser.nextToken();
-                ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.currentToken(), parser::getTokenLocation);
-            }
-            XContentParser.Token token = parser.currentToken();
-            String currentFieldName = null;
-            TotalHits totalHits = null;
-            ArrayList<SearchHit> searchHits = null;
-            ArrayList<Sequence> sequences = null;
-            ArrayList<Count> counts = null;
-
-            while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
-                if (token == XContentParser.Token.FIELD_NAME) {
-                    currentFieldName = parser.currentName();
-                } else if (token.isValue()) {
```

Comment:
I didn't implement this path since it existed in the original parser only for backward compatibility with the older version hits total and shouldn't occur in this version.